python/GafferUI/CodeWidget.py

```diff
@@ -0,0 +1,577 @@
+
+	def setCompleter( self, completer ) :
+
+		self.__completer = completer
+
+	def getCompleter( self ) :
+
+		return self.__completer
+
+	def setHighlighter( self, highlighter ) :
+
+		self.__highlighter.setHighlighter( highlighter )
+
+	def getHighlighter( self ) :
+
+		return self.__highlighter.getHighlighter()
+
+	def setCommentPrefix( self, commentPrefix ) :
+
+		self.__commentPrefix = commentPrefix
+
+	def getCommentPrefix( self ) :
+
+		return self.__commentPrefix
+
+	def __keyPress( self, widget, event ) :
+
+		if event.key == "Tab" :
+			if self._qtWidget().textCursor().hasSelection() :
+				if event.modifiers == event.Modifiers.None_ :
+					return self.__indentPress( 1 )
+				elif event.modifiers == event.Modifiers.Shift :
+					return self.__indentPress( -1 )
+			elif event.modifiers == event.Modifiers.None_ :
+				return self.__completionPress()
+		elif event.key == "Backtab" :
+			return self.__indentPress( -1 )
+		elif event.modifiers == event.Modifiers.Control and event.key == "BracketLeft" :
+			return self.__indentPress( -1 )
+		elif event.modifiers == event.Modifiers.Control and event.key == "BracketRight" :
+			return self.__indentPress( 1 )
+		elif event.modifiers == event.Modifiers.Control and event.key == "Slash" :
+			return self.__commentPress()
+		elif event.key == "Return" and event.modifiers == event.Modifiers.None_ :
+			return self.__returnPress()
+
+		return False
+
+	def __completionsAndLine( self ) :
+
+		if self.__completer is None :
+			return [], ""
+
+		text = self.getText()
+		cursor = self.getCursorPosition()
+		lineStart = max( text.rfind( "\n", 0, cursor ), 0 )
+		line = text[lineStart:cursor]
+
+		return self.__completer.completions( line ), line
+
+	def __completionPress( self ) :
+
+		completions, line = self.__completionsAndLine()
+		if not completions :
+			return False
+
+		commonPrefix = os.path.commonprefix( [ c.text for c in completions ] )
+		if len( commonPrefix ) > len( line ) :
+			self.insertText( commonPrefix[len(line):] )
+
+		return bool( completions )
+
+	def __selectedBlocks( self ) :
+
+		cursor = self._qtWidget().textCursor()
+		startBlock = self._qtWidget().document().findBlock( cursor.selectionStart() )
+		endBlock = self._qtWidget().document().findBlock( cursor.selectionEnd() )
+
+		result = []
+		while True :
+			result.append( startBlock )
+			if startBlock == endBlock :
+				break
+			else :
+				startBlock = startBlock.next()
+
+		if len( result ) > 1 and result[-1].position() == cursor.selectionEnd() :
+			# Omit last block if cursor is at the start of the line.
+			del result[-1]
+
+		return result
+
+	def __indentPress( self, indent ) :
+
+		cursor = self._qtWidget().textCursor()
+		try :
+
+			cursor.beginEditBlock()
+
+			for block in self.__selectedBlocks() :
+
+				cursor = QtGui.QTextCursor( block )
+				if indent > 0 :
+					cursor.insertText( "\t" * indent )
+				else :
+					for i in range( 0, len( os.path.commonprefix( [ block.text(), "\t" * -indent ] ) ) ) :
```

Comment:
I think maybe it would be something like `block.text()[cursor.blockPosition()]`. If we really want to revisit this though, we should probably change approach even further, and try to support the indent-with-spaces folks. I was hoping to avoid that for now at least - this was intended to be a quick cleanup of long-lost work and I've already spent a bunch of time adding new features when the HierarchyView requires my attention.